### ISSUE TITLE
Plugin 3rd party signature verification mechanism

### ISSUE THREAD

**Issue by wz2b** [grafana/grafana#26287]
**What would you like to be added**:
Add a way for 3rd parties to sign their own plugins

**Why is this needed**:
Grafana users would benefit by knowing and being able to verify that third party plugins they are using came from a trusted source.  Ideally, they would be able to trace that source.

**Ideas**
I have been thinking about this for a while, and searched the Issues (open and completed) to see if there was any discussion of ideas to implement code signing for 3rd parties.  I did not find a matching ticket, so I'm creating one as a place to capture a few ideas.  Documentation on code signing says this:

> Signing a plugin allows Grafana to verify the authenticity of the plugin with signature verification. This gives users a way to make sure plugins haven't been tampered with.

That's true, but if done properly signatures have another advantage: they give end users a verified path back to the developer - in other words, not just that it has been signed and not tampered with, but who did the signing.

***Via Public Key Infrasturcture***

For developers who have a CA-issued code signing certificate, it might not be a bad idea to have grafana accept code-signing certificates by simply trusting the well known root CAs in the public key infrastructure.

A key advantage of this is that anybody can look at the public certificate and know where the plugin came from.  If the author belongs to a well-known organization that a well-known CA recognizes that gives the administrator some confidence.  It doesn't necessarily say that the plugin is safe, but it does help establish that someone is accountable for it.

Keep in mind, though, that code signing certificates are pretty expensive (I have seen US$300/yr), so this should probably not be the _only_ way to sign plug-ins.  To keep encouraging open source development there probably needs to be another way.

***Self-signed plugins***

An approach to handling self-signed plugins is to allow administrators to install the public key of the developer they are trusting.  This would allow developers (especially open source developers) to use a self-signed certificate.  However, it's important that the certificate not be installable via the Grafana administrative user interface or any API.  A good pattern would be that 3rd party self-signed public certificates would have to be installed via the filesystem, via insertion in an application or system keyring (e.g. gnupg), or somehow enabled via custom.ini.  The key risk this mitigates is somebody getting hold of the grafana administrative password and installing a certificate that allows them to run arbitrary code.

This approach would be good for internal developers - those creating a plugin for a specific purpose within their own enterprise, without necessarily intending to release it to the public.

***Grafana becomes a CA***

Another possible approach is Grafana becomes its own CA (not a very authoritative one).  Third party developers could register with Grafana by sending a CSR.  Grafana could verify the users identity and sign the certificate for the developer.  Unfortunately, the burden of verification seems high.

**Sideloading plugins**

It occurs to me that it would be nice to load plug-ins via the UI, rather than having to sideload them by unzipping files in the right place, or worse, trying to use grafana-cli on a non-repository-hosted plugin.  Being able to upload plugins via the UI is potentially risky, but that risk is fairly mitigated (IMO) by having actual signatures on the plugins, provided there's no way to register or modify a new public certificate via that same interface.

In other words, I think it'd be OK to allow you to upload a .zip if everything is signed.

An idea on how to handle this would be to require you to use grafana-cli to install new public developer certificates.

**Comment by ivanahuckova** [comment]
fyi @ryantxu

**Comment by wz2b** [comment]
Something else occurred to me.  If Grafana is implementing a complete strategy to allow 3rd party signatures, it might also be useful to have some kind of installation-based or serial-number-based validation included.  This would allow 3rd parties to release their own 'enterprise' plugins on a per-user basis.  If you wanted to be fancy it could even include a time-limited 'trial period' mechanism.

**Comment by wbrowne** [comment]
@wz2b Grafana has a pretty mature signature verification mechanism by now - see docs [here](https://grafana.com/docs/grafana/latest/plugins/plugin-signatures/). Does this suit your needs? In the meantime I will close this issue, but we can re-open if we're missing something that should be supported.

**Comment by wz2b** [comment]
It doesn't exactly, no.  Public means you have to send the plugin to Grafana and they review it, and if it looks safe Grafana signs it and puts it in the public plugin repository.  Private means that you have to know exactly where you're going to install it.  There's some space in between that's not really well-addressed whereby you may want to have your own plugin but not have to list every single rootUrl where it will run.  In my use case it's not that these things are private or proprietary, they are just either 1) not of general use, so there's no good reason to put them in the catalog, or 2) in some sort of testing phase where I'd like to have a couple of people try it out without telling them they have to disable plugin verification entirely.

To avoid that, I think what I'd really like would be a way to load additional public certificates from grafana.ini.  To do this you only have to touch grafana.ini once and load a cert that says "Trust plugins signed by Chris."  Then, Grafana would just add that to the list of things it trusts.  In the public infrastructure mode, you'd load a well-known, trusted CA.  It's important you not be able to load new certs from the UI which is why I thought this through and decided grafana.ini is likely the place to do it.  Alternatively, if grafana had some kind of keystore (separate file) that you could manipulate that would be fine too.

So no ... the current certificate handling really does not do what I want it to do.  It's perfectly find to close this ticket and say "We don't want to do that" or "We don't see enough of a use case for that" but you can't quite say "It already does that with what we have" because it's not quite the same.

I'm on the grafana community slack if you want to discuss this more interactively.

**Comment by wbrowne** [comment]
>Private means that you have to know exactly where you're going to install it. There's some space in between that's not really well-addressed whereby you may want to have your own plugin but not have to list every single rootUrl where it will run.

[This issue](https://github.com/grafana/grafana/issues/3635) may be of interest to you in that case - something we plan to tackle soon.

>So no ... the current certificate handling really does not do what I want it to do. It's perfectly find to close this ticket and say "We don't want to do that" or "We don't see enough of a use case for that" but you can't quite say "It already does that with what we have" because it's not quite the same.

@wz2b your most recent message helped clarify the problem you're having and I agree that use-case is a little bit different. Let me chat with some folks to see if this is something that has come up or can be addressed in a different way

**Comment by wz2b** [comment]
Sounds good, yeah, allowing SPECIFIC unsigned plug-ins is definitely a good thing.  Maybe my additional use cases are too much on the edge to be valuable to Grafana users ovall.  I really can't speak to that.  From my perspective, there is just something in between "beta, experimental, and dangerous" and "in the grafana official repository" so to me it seemed one thing that could be done would be to use CA or individual certs to do something kind of like how you would add a PPA on an Ubuntu system.

Maybe that's too elaborate / more than you need.

Certs do have to come from _somewhere_ though ... implying to me that there's some kind of keystore somewhere.  Perhaps not, though - perhaps it's hard-coded.  I never looked at that part of the source.

**Comment by wbrowne** [comment]
>Maybe that's too elaborate / more than you need.

Time will tell! It's certainly an interesting idea and something we'll keep in the back of our minds.

>Certs do have to come from somewhere though ... implying to me that there's some kind of keystore somewhere. Perhaps not, though - perhaps it's hard-coded. I never looked at that part of the source.

https://github.com/grafana/grafana/blob/main/pkg/plugins/manager/manifest.go#L29

Another thing on our backlog! 😅
